sql: show create all types command

jackcwu opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/67685

Release note (sql change): SHOW CREATE ALL TYPES is a command that allows the user to get the statements to recreate
user-defined types of the current database. The command returns a flat log of the create statements for types.

Example output:
```
SHOW CREATE ALL TYPES
----
CREATE TYPE public.tableobj AS ENUM ('row', 'col');

```

RichardJCai reviewed (changes requested):
Nice thanks for getting this out quickly! 
Looks good overall just some minor comments.

> Reviewed 14 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-:-Mljzu5_5GB902DrA8eE:b-d3okgq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @jackcwu, and @rafiss)

___
*[-- commits, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljyzsB5DnUubgHqUoU):*
Nit: we can exclude enumerated here, just user-defined types is fine.
___
*[pkg/ccl/backupccl/targets.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljzFn17_4uI-0I44Pq:-MljzFn17_4uI-0I44Pr:b6zltbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/8761ef7513051479cab5f334a9aa35edabd36b52/pkg/ccl/backupccl/targets.go#L48)):*
> ```Go
> // represented as nil. Fills in the `priorIDs` map in the process, which maps
> // a descriptor the ID by which it was previously known (e.g pre-TRUNCATE).
> func getRelevantDescChanges(
> ```

Why did this file change in this commit?
___
*[pkg/sql/crdb_internal.go, line 2233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljzSt5AiFxSTMBGXSQ:-MljzSt5AiFxSTMBGXSR:b19a2cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9437d9677eb2750ee1fccd533130fd3981be79e/pkg/sql/crdb_internal.go#L2233)):*
> ```Go
>   create_statement   STRING,
>   enum_members       STRING[], -- populated only for ENUM types
> 	INDEX (descriptor_id)
> ```

Don't need to remove this


<!-- Sent from Reviewable.io -->

jackcwu committed:
sql: show create all types command

Release note (sql change): SHOW CREATE ALL TYPES is a command that allows the user to get the statements to recreate
user-defined types of the current database. The command returns a flat log of the create statements for types.

jackcwu reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71326)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @rafiss, and @RichardJCai)

___
*[-- commits, line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljyzsB5DnUubgHqUoU):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Nit: we can exclude enumerated here, just user-defined types is fine.
</blockquote></details>

fixed
___
*[pkg/sql/crdb_internal.go, line 2233 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljzSt5AiFxSTMBGXSQ:-MlovkXc4UtDHQWFJQ1-:by9dcu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9437d9677eb2750ee1fccd533130fd3981be79e/pkg/sql/crdb_internal.go#L2233)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Don't need to remove this
</blockquote></details>

put the enum_members, back in, removed the index like we discussed
___
*[pkg/ccl/backupccl/targets.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-MljzFn17_4uI-0I44Pq:-MlovpsM6kuf2IzRO5VC:b17sqcw) ([raw file](https://github.com/cockroachdb/cockroach/blob/8761ef7513051479cab5f334a9aa35edabd36b52/pkg/ccl/backupccl/targets.go#L48)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Why did this file change in this commit?
</blockquote></details>

used a new branch so these are no longer changed


<!-- Sent from Reviewable.io -->

jackcwu reviewed (commented):
fixed

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71326#-:-MlovwO_-1iUs-QdmQqg:b-7ql1td)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
Would be ready to merge after that.

RichardJCai reviewed (approved):
Nice! LGTM!

jackcwu commented:
bors r+
